versions/draft-3/SPEC.md

```diff
@@ -1515,6 +1522,130 @@ workflow wf {
 
+## Struct Definition
+A struct is a C-like construct which enables the user to create new compound types that consist of previously existing types. Structs
+can then be used within a `Task` or `Workflow` definition as a declaration in place of any other normal types. The struct takes the place of the
+`Object` type in many circumstances and enables proper typing of its members.
+
+Structs are defined using the `struct` keyword and have a single section consiting of typed declarationns. They are evaluated first prior
+to workflows and tasks.
+```wdl
+struct name { ... }
+```
+
+### Struct Declarations
+The only contents of struct are a set of declarations. Declarations can be any primitive or compound type, as well as other structs, and are defined
+the same way as they are in any other section. The one caveat to this is that declarations within a struct do not allow an expression statement after
+the initial member declaration.
+
+for example the following is a valid struct definition
+```wdl
+struct name {
+    String myString
+    Int myInt
+}
+```
+Whereas the following is invalid
+
+```wdl
+struct invalid {
+    String myString = "Cannot do this"
+    Int myInt
+}
+```
+
+You can also use compound types within a struct to easily encapsulate them within a single object. For example
+```wdl
+struct name {
+    Array[Array[File]] myFiles
+    Map[String,Pair[String,File]] myComplexType
+    String cohortName
+}
+```
+#### Optional and non Empty Struct Values
+Struct declarations can be optional or non-empty (if they are an array type). 
+
+```wdl
+struct name {
+    Arrray[File]+ myFiles
+    Boolean? myBoolean
+}
+```
+
+### Using a Struct
+When using a struct in the declaration section of either a `workflow` or a `task` or `output` section you define them in the same way you would define any other type.
+
+For example, if I have a struct like the following:
+```wdl
+struct MyStruct {
+    String myName
+    Int myAge
+}
+```
+
+then usage of the struct in a workflow would look like the following:
+
+```wdl
+
+task myTask {
+    MyStruct a
+    command {
+        echo "hello my name is ${a.myName} and I am ${a.myAge} years old"
+    }
+}
+
+workflow myWorkflow {
+    MyStruct a
+    call myTask {
+        input:
+            a = a
+    }
+
+}
+```
+
+#### Struct Assignment from Object Literal
+Structs can be assigned using an object literal. When Writing the object, all entries must conform or be coercible into the underlying type they are being assigned to
+
+```wdl
+
+MyStruct a = {"myName": "John","myAge": 30}
+
+```
+
+
+### Struct Member Access
+In order to access members within a struct, use object notation; ie `myStruct.myName`. If the uderlying member is a complex type which supports member access,
+you can access its elements in the way defined by that specific type.
+
+for example if we have a struct like the following:
+```wdl
+struct myStruct {
```

Comment:
use more different names to the example struct's, and perhaps keep to FirstCap in type names?

Reply:
I think "Types names must be `TitleCase`" is a good idea

Reply:
@yfarjoun will do for examples.
@cjllanwarne I talked to @geoffjentry I think, UpperCase first letter should be convention > enforced. WDYT?

Reply:
I tend to be in the "if it's a convention that everyone is expected to follow, it might as well be enforced" camp (eg Java class names might as well be enforced `UpperCase` at this point IMO).

I don't see a big cost to enforcing UpperCase. Every other type name is `UpperCase` so people are already keyed to the idea that that's what types look like.

Reply:
My $0.02 is that convention is fine unless it's problematic for defining in parsing/etc. 

I get the allusions to other languages but at the end of the day WDL is being marketed as a language for people who don't want to be using programming languages. If it just starts taking every trope from popular prog langs, why would users bother using it when they could just go over to e.g. python?